solidity_contracts/src/L1L2Messaging/L1KakarotMessaging.sol

```diff
@@ -0,0 +1,54 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.7.0 <0.9.0;
+
+import "../starknet/IStarknetMessaging.sol";
+
+interface IL1KakarotMessaging {
+    function sendMessageToL2(address to, uint248 value, bytes memory data) external payable;
+}
+
+contract L1KakarotMessaging  {
+
+    IStarknetMessaging private _starknetMessaging;
+    uint256 private _kakarotAddress;
+    uint256 constant HANDLE_L1_MESSAGE_SELECTOR = uint256(keccak256("handle_l1_message")) % 2**250;
+
+    constructor(address starknetMessaging, uint256 kakarotAddress) {
+        _starknetMessaging = IStarknetMessaging(starknetMessaging);
+        _kakarotAddress = kakarotAddress;
+
+    }
+
+    /// @notice Sends a message to a contract on L2.
+    /// @dev The bytes are split into individual uint256 values to use with the Starknet messaging system.
+    /// @dev This function must be called with a value sufficient to pay for the L1 message fee.
+    /// @param to The address of the contract on L2 to send the message to.
+    /// @param value The value to send to the contract on L2. The value is taken from the L2 contract address.
+    /// @param data The data to send to the contract on L2.
+    function sendMessageToL2(address to, uint248 value, bytes calldata data) external payable {
+        uint256[] memory convertedData = new uint256[](data.length + 4);
+        convertedData[0] = uint256(uint160(msg.sender));
+        convertedData[1] = uint256(uint160(to));
+        convertedData[2] = uint256(value);
+        convertedData[3] = uint256(data.length);
+        for (uint256 i = 4; i < data.length + 4; i++) {
+            convertedData[i] = uint256(uint8(data[i - 4]));
+        }
+
+        // Send the converted data to L2
+        _starknetMessaging.sendMessageToL2{value: msg.value}(
+            _kakarotAddress,
+            HANDLE_L1_MESSAGE_SELECTOR,
+            convertedData
+        );
+    }
+
+    /// @notice Consumes a message sent from L2.
+    /// @dev Must be called with a delegatecall so that the `msg.sender` in the StarknetMessaging contract
+    ///     is the caller of this function.
+    /// @param payload The payload of the message to consume.
+    function consumeMessageFromL2(uint256[] calldata payload) external returns (bytes32 msgHash){
```

Comment:
The signature must be consistent with the sendMessageToL1

```
    function sendMessageToL1(
        address toAddress,
        uint248[] memory data
    ) internal {
```

So
1. decide wether we use `uint256` or `uint248`
2. wether it's `data` or `payload`

Reply:
as per previous comment, we can have 2 entrypoints here, one with `uint256[] data` where the user packs off-chain to mimic the behavior of the precompile, and another with `bytes memory data` where we pack in yul

Reply:
if we don't pack:
- sends `b"1234"` on L2
- calls `consumeMessageFromL2` with `list(b"1234")` on L1

if we pack:
- sends `b"1234"` on L2
- calls `consumeMessageFromL2` with `[int(w.ljust(2 * 31, "0"), 16) for w in wrap(b"1234".hex(), 31)]`

Reply:
Let's keep packing for another PR ?